src/Microsoft.AspNetCore.Http.Connections/Internal/HttpConnectionContext.cs

```diff
@@ -180,7 +180,7 @@ public void TickHeartbeat()
         {
             var disposeTask = Task.CompletedTask;
 
-            await StateLock.WaitAsync();
+            await Task.WhenAll(StateLock.WaitAsync(), WriteLock.WaitAsync());
```

Comment:
There was a reason we didn't take both locks here
I think this might deadlock if there is backpressure at https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.Http.Connections/Internal/HttpConnectionDispatcher.cs#L482

Reply:
Don't we have a test for this?

Reply:
So the issue is that because of back pressure the writer will wait until it can write again? Are you worried that we don't want the dispose to hang around, waiting on the write lock to become available?

Would a solution be to have a cancellation token that the connection is closing. That will break users of the connection out of waiting on locks and allow a more graceful shutdown.

Reply:
We should call CancelPendingFlush to yield the backpressure (not sure why we aren’t doing this already).

Does this mean we can go back to a single lock? I need to think about this a bit

Reply:
I was only talking about the implications of the dispose scenario if both locks were taken in it. CancelPendingFlush might help there.